File: libs/trading_models/signals.py
"""Trading signal models."""

from datetime import datetime
from decimal import Decimal
from typing import Any, Optional

from pydantic import Field, model_validator

from .base import BaseModel
from .enums import Direction, MarketRegime, Timeframe
from .market_data import IndicatorSnapshot
from .patterns import PatternHit


class TimeframeAnalysis(BaseModel):
    """Analysis results for a specific timeframe."""

    timeframe: Timeframe = Field(..., description="Analysis timeframe")
    timestamp: datetime = Field(..., description="Analysis timestamp")

    # Technical analysis scores
    trend_score: float = Field(..., ge=-10, le=10, description="Trend strength (-10 to 10)")
    momentum_score: float = Field(..., ge=-10, le=10, description="Momentum score (-10 to 10)")
    volatility_score: float = Field(..., ge=0, le=10, description="Volatility score (0-10)")
    volume_score: float = Field(..., ge=0, le=10, description="Volume score (0-10)")

    # Pattern analysis
    pattern_count: int = Field(0, ge=0, description="Number of patterns detected")
    strongest_pattern_confidence: float = Field(0, ge=0, le=1, description="Strongest pattern confidence")

    # Indicator summary
    bullish_indicators: int = Field(0, ge=0, description="Count of bullish indicators")
    bearish_indicators: int = Field(0, ge=0, description="Count of bearish indicators")
    neutral_indicators: int = Field(0, ge=0, description="Count of neutral indicators")

    # Weight for confluence calculation
    timeframe_weight: float = Field(..., gt=0, le=1, description="Weight in confluence calculation")


class LLMAnalysis(BaseModel):
    """LLM analysis results."""

    model_id: str = Field(..., description="LLM model identifier")
    timestamp: datetime = Field(..., description="Analysis timestamp")

    # Analysis results
    market_sentiment: str = Field(..., description="Market sentiment analysis")
    key_insights: list[str] = Field(default_factory=list, description="Key market insights")
    risk_factors: list[str] = Field(default_factory=list, description="Identified risk factors")

    # Scoring
    bullish_score: float = Field(..., ge=0, le=10, description="Bullish sentiment score (0-10)")
    bearish_score: float = Field(..., ge=0, le=10, description="Bearish sentiment score (0-10)")
    confidence: float = Field(..., ge=0, le=1, description="Analysis confidence (0-1)")

    # Performance metrics
    tokens_used: int = Field(..., gt=0, description="Tokens consumed")
    latency_ms: int = Field(..., gt=0, description="Response latency in milliseconds")
    cost_usd: float = Field(..., ge=0, description="Analysis cost in USD")


class Signal(BaseModel):
    """Trading signal with confluence analysis."""

    signal_id: str = Field(..., description="Unique signal identifier")
    symbol: str = Field(..., description="Trading symbol")
    timestamp: datetime = Field(..., description="Signal generation timestamp")

    # Signal direction and strength
    direction: Direction = Field(..., description="Signal direction (long/short)")
    confluence_score: float = Field(..., ge=0, le=100, description="Confluence score (0-100)")
    confidence: float = Field(..., ge=0, le=1, description="Signal confidence (0-1)")

    # Market context
    market_regime: MarketRegime = Field(..., description="Current market regime")
    primary_timeframe: Timeframe = Field(..., description="Primary analysis timeframe")

    # Analysis components
    timeframe_analysis: dict[Timeframe, TimeframeAnalysis] = Field(
        default_factory=dict,
        description="Analysis by timeframe"
    )
    patterns: list[PatternHit] = Field(default_factory=list, description="Supporting patterns")
    indicators: dict[Timeframe, IndicatorSnapshot] = Field(
        default_factory=dict,
        description="Indicator snapshots by timeframe"
    )
    llm_analysis: Optional[LLMAnalysis] = Field(None, description="LLM analysis results")

    # Price targets
    entry_price: Optional[Decimal] = Field(None, gt=0, description="Suggested entry price")
    stop_loss: Optional[Decimal] = Field(None, gt=0, description="Suggested stop loss")
    take_profit: Optional[Decimal] = Field(None, gt=0, description="Suggested take profit")

    # Risk metrics
    risk_reward_ratio: Optional[float] = Field(None, gt=0, description="Risk/reward ratio")
    max_risk_pct: Optional[float] = Field(None, gt=0, le=10, description="Maximum risk percentage")

    # Signal reasoning
    reasoning: str = Field(..., description="Human-readable signal reasoning")
    key_factors: list[str] = Field(default_factory=list, description="Key supporting factors")

    # Metadata
    expires_at: Optional[datetime] = Field(None, description="Signal expiration time")
    priority: int = Field(1, ge=1, le=5, description="Signal priority (1-5)")

    @model_validator(mode='after')
    def validate_signal_consistency(self):
        """Validate signal consistency."""
        # High confluence should have high confidence
        if self.confluence_score > 90 and self.confidence < 0.8:
            raise ValueError("High confluence score requires high confidence")

        # Validate risk/reward ratio calculation
        if (self.risk_reward_ratio and self.entry_price and
            self.stop_loss and self.take_profit):

            if self.direction == Direction.LONG:
                risk = float(self.entry_price - self.stop_loss)
                reward = float(self.take_profit - self.entry_price)
            else:  # SHORT
                risk = float(self.stop_loss - self.entry_price)
                reward = float(self.entry_price - self.take_profit)

            if risk > 0:
                calculated_rr = reward / risk
                if abs(calculated_rr - self.risk_reward_ratio) > 0.1:
                    raise ValueError("Risk/reward ratio doesn't match price levels")

        return self

    def add_timeframe_analysis(self, analysis: TimeframeAnalysis) -> None:
        """Add timeframe analysis to the signal."""
        self.timeframe_analysis[analysis.timeframe] = analysis

    def get_weighted_confluence(self) -> float:
        """Calculate weighted confluence score from timeframe analyses."""
        if not self.timeframe_analysis:
            return self.confluence_score

        total_weight = sum(ta.timeframe_weight for ta in self.timeframe_analysis.values())
        if total_weight == 0:
            return self.confluence_score

        weighted_sum = sum(
            (ta.trend_score + ta.momentum_score) * ta.timeframe_weight
            for ta in self.timeframe_analysis.values()
        )

        # Normalize to 0-100 scale
        normalized_score = max(0, min(100, (weighted_sum / total_weight + 10) * 5))
        return normalized_score


class TradingSignal(BaseModel):
    """Simplified trading signal for backtesting and testing."""

    symbol: str = Field(..., description="Trading symbol")
    direction: Direction = Field(..., description="Trade direction")
    confidence: float = Field(..., ge=0, le=1, description="Signal confidence")
    position_size: float = Field(..., gt=0, description="Position size")
    stop_loss: Optional[float] = Field(None, description="Stop loss price")
    take_profit: Optional[float] = Field(None, description="Take profit price")
    reasoning: str = Field(..., description="Signal reasoning")
    timeframe_analysis: dict[str, Any] = Field(default_factory=dict, description="Timeframe analysis")

File: libs/rust-common/src/trading_models/signals.rs
//! Trading signal structures.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use super::{
    enums::{Direction, MarketRegime, Timeframe},
    market_data::IndicatorSnapshot,
    patterns::PatternHit,
};

/// Analysis results for a specific timeframe.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TimeframeAnalysis {
    pub timeframe: Timeframe,
    pub timestamp: DateTime<Utc>,
    
    // Technical analysis scores
    pub trend_score: f64,
    pub momentum_score: f64,
    pub volatility_score: f64,
    pub volume_score: f64,
    
    // Pattern analysis
    pub pattern_count: u32,
    pub strongest_pattern_confidence: f64,
    
    // Indicator summary
    pub bullish_indicators: u32,
    pub bearish_indicators: u32,
    pub neutral_indicators: u32,
    
    // Weight for confluence calculation
    pub timeframe_weight: f64,
}

impl TimeframeAnalysis {
    /// Validate analysis scores.
    pub fn validate(&self) -> Result<(), String> {
        // Validate score ranges
        if !(-10.0..=10.0).contains(&self.trend_score) {
            return Err("Trend score must be between -10 and 10".to_string());
        }
        
        if !(-10.0..=10.0).contains(&self.momentum_score) {
            return Err("Momentum score must be between -10 and 10".to_string());
        }
        
        if !(0.0..=10.0).contains(&self.volatility_score) {
            return Err("Volatility score must be between 0 and 10".to_string());
        }
        
        if !(0.0..=10.0).contains(&self.volume_score) {
            return Err("Volume score must be between 0 and 10".to_string());
        }
        
        // Validate confidence range
        if !(0.0..=1.0).contains(&self.strongest_pattern_confidence) {
            return Err("Pattern confidence must be between 0 and 1".to_string());
        }
        
        // Validate timeframe weight
        if !(0.0..=1.0).contains(&self.timeframe_weight) || self.timeframe_weight == 0.0 {
            return Err("Timeframe weight must be between 0 (exclusive) and 1".to_string());
        }
        
        Ok(())
    }
}

/// LLM analysis results.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LlmAnalysis {
    pub model_id: String,
    pub timestamp: DateTime<Utc>,
    
    // Analysis results
    pub market_sentiment: String,
    pub key_insights: Vec<String>,
    pub risk_factors: Vec<String>,
    
    // Scoring
    pub bullish_score: f64,
    pub bearish_score: f64,
    pub confidence: f64,
    
    // Performance metrics
    pub tokens_used: u32,
    pub latency_ms: u32,
    pub cost_usd: f64,
}

impl LlmAnalysis {
    /// Validate LLM analysis data.
    pub fn validate(&self) -> Result<(), String> {
        // Validate score ranges
        if !(0.0..=10.0).contains(&self.bullish_score) {
            return Err("Bullish score must be between 0 and 10".to_string());
        }
        
        if !(0.0..=10.0).contains(&self.bearish_score) {
            return Err("Bearish score must be between 0 and 10".to_string());
        }
        
        if !(0.0..=1.0).contains(&self.confidence) {
            return Err("Confidence must be between 0 and 1".to_string());
        }
        
        // Validate performance metrics
        if self.tokens_used == 0 {
            return Err("Tokens used must be greater than 0".to_string());
        }
        
        if self.latency_ms == 0 {
            return Err("Latency must be greater than 0".to_string());
        }
        
        if self.cost_usd < 0.0 {
            return Err("Cost must be non-negative".to_string());
        }
        
        Ok(())
    }
}

/// Trading signal with confluence analysis.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Signal {
    pub signal_id: String,
    pub symbol: String,
    pub timestamp: DateTime<Utc>,
    
    // Signal direction and strength
    pub direction: Direction,
    pub confluence_score: f64,
    pub confidence: f64,
    
    // Market context
    pub market_regime: MarketRegime,
    pub primary_timeframe: Timeframe,
    
    // Analysis components
    pub timeframe_analysis: HashMap<Timeframe, TimeframeAnalysis>,
    pub patterns: Vec<PatternHit>,
    pub indicators: HashMap<Timeframe, IndicatorSnapshot>,
    pub llm_analysis: Option<LlmAnalysis>,
    
    // Price targets
    pub entry_price: Option<f64>,
    pub stop_loss: Option<f64>,
    pub take_profit: Option<f64>,
    
    // Risk metrics
    pub risk_reward_ratio: Option<f64>,
    pub max_risk_pct: Option<f64>,
    
    // Signal reasoning
    pub reasoning: String,
    pub key_factors: Vec<String>,
    
    // Metadata
    pub expires_at: Option<DateTime<Utc>>,
    pub priority: u8,
}

impl Signal {
    /// Validate signal data.
    pub fn validate(&self) -> Result<(), String> {
        // Validate confluence score
        if !(0.0..=100.0).contains(&self.confluence_score) {
            return Err("Confluence score must be between 0 and 100".to_string());
        }
        
        // Validate confidence
        if !(0.0..=1.0).contains(&self.confidence) {
            return Err("Confidence must be between 0 and 1".to_string());
        }
        
        // High confluence should have high confidence
        if self.confluence_score > 90.0 && self.confidence < 0.8 {
            return Err("High confluence score requires high confidence".to_string());
        }
        
        // Validate price levels
        if let Some(entry) = self.entry_price {
            if entry <= 0.0 {
                return Err("Entry price must be positive".to_string());
            }
        }
        
        if let Some(stop) = self.stop_loss {
            if stop <= 0.0 {
                return Err("Stop loss must be positive".to_string());
            }
        }
        
        if let Some(target) = self.take_profit {
            if target <= 0.0 {
                return Err("Take profit must be positive".to_string());
            }
        }
        
        // Validate risk/reward ratio
        if let Some(rr) = self.risk_reward_ratio {
            if rr <= 0.0 {
                return Err("Risk/reward ratio must be positive".to_string());
            }
            
            // Validate against price levels if available
            if let (Some(entry), Some(stop), Some(target)) = (self.entry_price, self.stop_loss, self.take_profit) {
                let (risk, reward) = match self.direction {
                    Direction::Long => (entry - stop, target - entry),
                    Direction::Short => (stop - entry, entry - target),
                };
                
                if risk > 0.0 {
                    let calculated_rr = reward / risk;
                    if (calculated_rr - rr).abs() > 0.1 {
                        return Err("Risk/reward ratio doesn't match price levels".to_string());
                    }
                }
            }
        }
        
        // Validate max risk percentage
        if let Some(max_risk) = self.max_risk_pct {
            if !(0.0..=10.0).contains(&max_risk) {
                return Err("Max risk percentage must be between 0 and 10".to_string());
            }
        }
        
        // Validate priority
        if !(1..=5).contains(&self.priority) {
            return Err("Priority must be between 1 and 5".to_string());
        }
        
        // Validate timeframe analyses
        for analysis in self.timeframe_analysis.values() {
            analysis.validate()?;
        }
        
        // Validate patterns
        for pattern in &self.patterns {
            pattern.validate()?;
        }
        
        // Validate indicators
        for indicator in self.indicators.values() {
            indicator.validate()?;
        }
        
        // Validate LLM analysis if present
        if let Some(llm) = &self.llm_analysis {
            llm.validate()?;
        }
        
        Ok(())
    }
    
    /// Calculate weighted confluence score from timeframe analyses.
    pub fn get_weighted_confluence(&self) -> f64 {
        if self.timeframe_analysis.is_empty() {
            return self.confluence_score;
        }
        
        let total_weight: f64 = self.timeframe_analysis.values()
            .map(|ta| ta.timeframe_weight)
            .sum();
        
        if total_weight == 0.0 {
            return self.confluence_score;
        }
        
        let weighted_sum: f64 = self.timeframe_analysis.values()
            .map(|ta| (ta.trend_score + ta.momentum_score) * ta.timeframe_weight)
            .sum();
        
        // Normalize to 0-100 scale
        let normalized_score = (weighted_sum / total_weight + 10.0) * 5.0;
        normalized_score.max(0.0).min(100.0)
    }
}